core/state/intra_block_state.go

```diff
@@ -321,24 +329,27 @@ func (sdb *IntraBlockState) AddBalance(addr libcommon.Address, amount *uint256.I
 	if sdb.trace {
 		fmt.Printf("AddBalance %x, %d\n", addr, amount)
 	}
-	// If this account has not been read, add to the balance increment map
-	_, needAccount := sdb.stateObjects[addr]
-	if !needAccount && addr == ripemd && amount.IsZero() {
-		needAccount = true
-	}
-	if !needAccount {
-		sdb.journal.append(balanceIncrease{
-			account:  &addr,
-			increase: *amount,
-		})
-		bi, ok := sdb.balanceInc[addr]
-		if !ok {
-			bi = &BalanceIncrease{}
-			sdb.balanceInc[addr] = bi
+
+	if sdb.logger == nil {
```

Comment:
Got it. I will trigger the tracer event at this level then.

Reply:
I looked into this and as per tracer implementation, we are required to report the OnBalanceChange that takes prev and new balance. In order to report this correctly and timely, we have to get the stateObject and log it. Kindly provide your guidance.
cc: @maoueh

Reply:
That's an interesting problem.

I've read the implementation of the various usage of `balanceInc` and it seems an optimization by Erigon to avoid/delay reading state object from DB on balance increase. As long as the state object is not read, the balance increase is deferred until the very end of the transaction (or block depending on execution context).

The tracing however wants to record balance change at the moment they are actually performed so that proper balance tracking can be done within a transaction and most importantly, the the overall "ordering" of what is happening within the transaction is kept relatively against other tracked state changes (nonce, code, etc).

Now, this is side-stepped by this optimization, a different proposition could be:

```diff
diff --git a/core/state/intra_block_state.go b/core/state/intra_block_state.go
index ff0574b2c5..9339b0d688 100644
--- a/core/state/intra_block_state.go
+++ b/core/state/intra_block_state.go
@@ -326,30 +326,36 @@ func (sdb *IntraBlockState) AddBalance(addr libcommon.Address, amount *uint256.I
        if !needAccount && addr == ripemd && amount.IsZero() {
                needAccount = true
        }
        if !needAccount {
                sdb.journal.append(balanceIncrease{
                        account:  &addr,
                        increase: *amount,
                })
                bi, ok := sdb.balanceInc[addr]
                if !ok {
                        bi = &BalanceIncrease{}
                        sdb.balanceInc[addr] = bi
                }
                bi.increase.Add(&bi.increase, amount)
                bi.count++
+
+               if sdb.logger != nil && sdb.logger.OnBalanceChange != nil {
+                       balance := sdb.GetOrNewStateObject(addr).Balance()
+                       sdb.logger.OnBalanceChange(addr, balance, new(uint256.Int).Add(balance, amount), reason)
+               }
+
                return
        }

        stateObject := sdb.GetOrNewStateObject(addr)
        stateObject.AddBalance(amount, reason)
 }
```

Ultimately, that has the same net effect as the current diff but written differently. Indeed, now that we call `GetOrNewStateObject` to get the current balance, we are going to populate `stateObjects[addr]` entry as well as transferring the balance that exist now. And since now the `addr` is populated in `stateObjects`, the next balance increase for this same account will not go in the `if !needAccount` statement anymore.

This diff above the same net effect as the current PR both skips the optimization altogether by the fact that we need to read the account and get the actual balance.

However, I think the proposed diff above is less obvious about the consequences of calling `GetOrNewStateObject` which modifies the actual state.

I personally feel that the first version with an appropriate comment would best reflect the intent of the change:

```diff
diff --git a/core/state/intra_block_state.go b/core/state/intra_block_state.go
index ff0574b2c5..2802bc6642 100644
--- a/core/state/intra_block_state.go
+++ b/core/state/intra_block_state.go
@@ -321,24 +321,32 @@ func (sdb *IntraBlockState) AddBalance(addr libcommon.Address, amount *uint256.I
        if sdb.trace {
                fmt.Printf("AddBalance %x, %d\n", addr, amount)
        }
-       // If this account has not been read, add to the balance increment map
-       _, needAccount := sdb.stateObjects[addr]
-       if !needAccount && addr == ripemd && amount.IsZero() {
-               needAccount = true
-       }
-       if !needAccount {
-               sdb.journal.append(balanceIncrease{
-                       account:  &addr,
-                       increase: *amount,
-               })
-               bi, ok := sdb.balanceInc[addr]
-               if !ok {
-                       bi = &BalanceIncrease{}
-                       sdb.balanceInc[addr] = bi
+
+       // If tracing of balance changes is enabled, we cannot apply optimization
+       // of not reading the account before increasing the balance. Indeed, the tracing
+       // requires knowledge of the balance prior the change and most importantly
+       // must record the change at the time it's applied.
+       tracingBalanceChange := sdb.logger != nil && sdb.logger.OnBalanceChange != nil
+       if !tracingBalanceChange {
+               // If this account has not been read, add to the balance increment map
+               _, needAccount := sdb.stateObjects[addr]
+               if !needAccount && addr == ripemd && amount.IsZero() {
+                       needAccount = true
+               }
+               if !needAccount {
+                       sdb.journal.append(balanceIncrease{
+                               account:  &addr,
+                               increase: *amount,
+                       })
+                       bi, ok := sdb.balanceInc[addr]
+                       if !ok {
+                               bi = &BalanceIncrease{}
+                               sdb.balanceInc[addr] = bi
+                       }
+                       bi.increase.Add(&bi.increase, amount)
+                       bi.count++
+                       return
                }
-               bi.increase.Add(&bi.increase, amount)
-               bi.count++
-               return
        }

        stateObject := sdb.GetOrNewStateObject(addr)
```

@AskAlexSharov Let us know your opinion here.

Reply:
balanceInc - it’s not optimisation of execution/state_reading. It’s reduction of “conflict rate in parallel-execution”.

let me explain: if run txs in-parallel - if 2 txs update same key X - it’s conflict. But if 2 txs IncreaseBalance of same account (without reading it) - it’s actually conflict-free operation.

So: balanceInc it’s “conflict-free updates” - this is reason why it’s stored separately from other updates.  

we don’t have working prototype of parallel execution now, but we had 1 year ago and we plan to have it in future. (FYI: conflict-resolution is done by ReadsValid func).

Reply:
Thanks for the explanation. @AskAlexSharov Do you see a way to this tracer requirement without hampering the existing implementation?

Don't we have the same conflict issue in the `SubBalance` operation?

Reply:
I have idea: we can add to `type StateReader interface {` method `ReadAccountDataForDebug` and use it only inside `if !needAccount` and if logger set. Example implementation: https://github.com/ledgerwatch/erigon/pull/11089/files


`SubBalance` - probably they are not conflict-free (may end up below zero). Or we just didn't implement it.

Reply:
@AskAlexSharov About `ReadAccountDataForDebug` from the PR you linked, this wouldn't account for `balanceIncrease` journaled no? We would need to correctly "track" those intermediate changes but it would be doable.

@dhyaniarun1993 Since you had change this code, I imagine this was resurfaced as a difference in battlefield test suite, is that correct? It would be good if that would be the case, since we will be able to test that new implementation. Are you good to take at look at what is would look like?

Reply:
Just instead of getObject (to get prev balance) - call ReadAccForDebug. In my head - it must just work.

Reply:
> Also let’s create geth issue about it. Theoretically it can be consensus bug or future bug (hooks must not affect access_list, etc…).

Not sure I understand, Geth doesn't have that optimization, and 100%, hooks is only meant to read and not modify anything, that's why I didn't like the `GetOrNewStateObject` proposition I did, I 100% prefer something that don't change any current state.

> Just instead of getObject (to get prev balance) - call ReadAccForDebug. In my head - it must just work.

What I see as a problem is if there is multiple transfer to an account and this account is never read:

```
// First increase, !needAccount
if logger {
  balance := ReadAccountBalanceDebug()
  // Record change
}

// Second increase, !needAccount
if logger {
  balance := ReadAccountBalanceDebug()
  // Will yield same balance value as first call, because didn't take into account first increase
}
```

Quite feasible to manage, just that calling straight `ReadAccountBalanceDebug` will not be enough.

Reply:
`Geth doesn't have that optimization` - from my perspective it's not just optimization - Hook's API forcing app to read different keys (which was not read before). And if you will include read_list in consensus or push user to pay for each Read, or allow user to pay less if key is in Cache (we read it recentely) - then it will end up with sub-optimal costs (reducible by just changing Hooks API). 

Multiple `balanceInc` - yes maybe this will be a problem for you. maybe you can add UnitTest for this case. Probably solvable by lookup inside `balanceInc     map[libcommon.Address]*BalanceIncrease`: `prev = sum(balanceInc[acc]) + `ReadAccountBalanceDebug().Balance`. Let's try do couple more steps in this direction - just to see where we come / is it possible or not.

Reply:
`Geth doesn't have that optimization` - maybe it can be a proposal - to add new hook method for balance increases events. Maybe it will be useful for some users... maybe it will open some new doors... 🚪

Reply:
@maoueh I don't remember where I saw this difference. I don't think it was recorded by battlefield.

I will work on the proposition and see where it leads us.

Reply:
I can see that we have `readList` only for `StateReaderV3`. What about other readers?

Reply:
> from my perspective it's not just optimization - Hook's API forcing app to read different keys (which was not read before)

Agreed, but only special node will pay this price, and there is no way around, someone will need to read at once to get the actual value before the increase.

But I'm all with you that if possible, we should avoid that, like @dhyaniarun1993 mentioned, we are going to explore the path you suggested, I'm confident we can make it work properly.

> I will work on the proposition and see where it leads us.

Ok let me know, I think that a double transfer to the same address in the same transaction will yield 2 `balanceInc` being generated. Let me know your findings, I can take a quick look at adding it in battlefield too, shouldn't be too hard and I could add a quick `println` call to ensure we correctly exercise all code paths.

The best would be to have an actual unit test that proves the correct tracing instrumentation. Have we ported over the "integration test" we had in Firehose Geth? (https://github.com/streamingfast/go-ethereum/blob/firehose-fh3.0/eth/tracers/internal/tracetest/firehose/firehose_test.go#L72)

Reply:
> I think that a double transfer to the same address in the same transaction will yield 2 balanceInc being generated

Do you mean in the tracer(`OnBalanceChange`) or the journal(`balanceIncrease`)?

> Have we ported over the "integration test" we had in Firehose Geth?

Yes. but that PR is not upto date with the current changes.

Reply:
Hey @AskAlexSharov 
Can you help me with [this](https://github.com/erigontech/erigon/issues/11462)?